crates/store/blobs/src/lib.rs

```diff
@@ -0,0 +1,213 @@
+        E: Into<eyre::Report>,
+    {
+        let chunks = typed_stream::<eyre::Result<_>>(try_stream!({
+            pin_mut!(stream);
+
+            // todo! use a bufreader
+            while let Some(blob) = stream.try_next().await? {
+                let blob = blob.as_ref();
+
+                for chunk in blob.chunks(CHUNK_SIZE) {
+                    let id = BlobId::hash(&chunk);
+
+                    self.data_store.handle().put(
+                        &calimero_store::key::BlobMeta::new(id),
+                        &calimero_store::types::BlobMeta {
+                            size: 0,
+                            links: Vec::new().into_boxed_slice(),
+                        },
+                    )?;
+                    self.blob_store.put(id, chunk).await?;
+
+                    yield id;
+                }
+
+                // let mut chunks = blob
+                //     .chunks(CHUNK_SIZE)
+                //     .map(|chunk| self.blob_store.put(BlobId::hash(&chunk), chunk))
+                //     .collect::<FuturesUnordered<_>>();
+
+                // while let Some(_) = chunks.try_next().await? {}
+            }
+        }));
+
+        // let ids = ids.chunks(MAX_LINKS_PER_BLOB);
+        pin_mut!(chunks);
+
+        let mut links = Vec::new();
+        let mut digest = sha2::Sha256::new();
+
+        while let Some(id) = chunks.try_next().await? {
+            links.push(calimero_store::key::BlobMeta::new(id));
+            digest.update(id.as_ref());
+        }
+
+        let id = BlobId::from(*(AsRef::<[u8; 32]>::as_ref(&digest.finalize())));
+
+        self.data_store.handle().put(
+            &calimero_store::key::BlobMeta::new(id),
+            &calimero_store::types::BlobMeta {
+                size: 0,
+                links: links.into_boxed_slice(),
+                // todo! hash of the blob data
+            },
+        )?;
+
+        Ok(id) // todo!: Ok((id, Blob { size, hash }::{fn stream()}))
+    }
+}
+
+fn typed_stream<T>(s: impl Stream<Item = T>) -> impl Stream<Item = T> {
+    s
+}
+
+pub struct Blob {
+    // id: BlobId,
+    // meta: calimero_store::types::BlobMeta,
+
+    // blob_mgr: BlobManager,
+    stream: Pin<Box<dyn Stream<Item = Result<Box<[u8]>, BlobError>>>>,
+}
+
+impl Blob {
+    async fn new(id: BlobId, blob_mgr: BlobManager) -> eyre::Result<Option<Self>> {
+        let Some(blob_meta) = blob_mgr
+            .data_store
+            .handle()
+            .get(&calimero_store::key::BlobMeta::new(id))?
+        else {
+            return Ok(None);
+        };
+
+        let stream = Box::pin(try_stream!({
+            if blob_meta.links.is_empty() {
+                let maybe_blob = blob_mgr.blob_store.get(id).await;
+                let maybe_blob = maybe_blob.map_err(BlobError::RepoError)?;
+                let blob = maybe_blob.ok_or_else(|| BlobError::DanglingBlob { id })?;
+                return yield blob;
+            }
+
+            for link in blob_meta.links {
```

Comment:
```suggestion
           for link in &*blob_meta.links  {
```

as [calimero_store::key::BlobMeta]` is not an iterator

Reply:
https://doc.rust-lang.org/std/boxed/struct.Box.html#impl-IntoIterator-for-Box%3C%5BI%5D,+A%3E

`Box<[T]>: IntoIterator<Item = T>` seems to have been stabilized in rust `1.80`, you need to update rust

![CleanShot 2024-08-06 at 11 24 20@2x](https://github.com/user-attachments/assets/48981d71-7eeb-4cf3-8ae8-74e7c378ac28)

cc: https://github.com/calimero-network/core/issues/433

Reply:
Ok, it wasn't noted. We have minimum requirements in the docs for this kind of cases: https://calimero-network.github.io/getting-started/setup

Currently, it is 1.79. I'll try with 1.80

Reply:
https://github.com/calimero-network/calimero-network.github.io/pull/11